docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+    # The following default values are invalid and the types are unspecified.
+    def invalid(a: int = "", b: Foo = Foo()): ...
 
-While stub files are syntactically valid Python modules, they use the
-``.pyi`` extension to make it possible to maintain stub files in the
-same directory as the corresponding real module.  This also reinforces
-the notion that no runtime behavior should be expected of stub files.
+For a class ``C``, the type of the first argument to a classmethod is
+assumed to be ``type[C]``, if unannotated. For other non-static methods,
+its type is assumed to be ``C``::
 
-Additional notes on stub files:
+    class Foo:
+        def do_things(self): ...  # self has type Foo
+        @classmethod
+        def create_it(cls): ...  # cls has type type[Foo]
+        @staticmethod
+        def utility(x): ...  # x has type Any
 
-* Modules and variables imported into the stub are not considered
-  exported from the stub unless the import uses the ``import ... as
-  ...`` form or the equivalent ``from ... import ... as ...`` form.
-  (*UPDATE:* To clarify, the intention here is that only names
-  imported using the form ``X as X`` will be exported, i.e. the name
-  before and after ``as`` must be the same.)
+But::
 
-* However, as an exception to the previous bullet, all objects
-  imported into a stub using ``from ... import *`` are considered
-  exported.  (This makes it easier to re-export all objects from a
-  given module that may vary by Python version.)
+    _T = TypeVar("_T")
 
-* Just like in `normal Python files <https://docs.python.org/3/reference/import.html#submodules>`_, submodules
-  automatically become exported attributes of their parent module
-  when imported. For example, if the ``spam`` package has the
-  following directory structure::
+    class Foo:
+        def do_things(self: _T) -> _T: ...  # self has type _T
+        @classmethod
+        def create_it(cls: _T) -> _T: ...  # cls has type _T
 
-      spam/
-          __init__.pyi
-          ham.pyi
+Using a function or method body other than the ellipsis literal is currently
+unspecified. Stub authors may experiment with other bodies, but it is up to
+individual type checkers how to interpret them::
 
-  where ``__init__.pyi`` contains a line such as ``from . import ham``
-  or ``from .ham import Ham``, then ``ham`` is an exported attribute
-  of ``spam``.
+    def foo(): ...  # compatible
+    def bar(): pass  # behavior undefined
 
-* Stub files may be incomplete. To make type checkers aware of this, the file
-  can contain the following code::
+All variants of overloaded functions and methods must have an ``@overload``
+decorator::
 
-    def __getattr__(name) -> Any: ...
+    @overload
+    def foo(x: str) -> str: ...
+    @overload
+    def foo(x: float) -> int: ...
 
-  Any identifier not defined in the stub is therefore assumed to be of type
-  ``Any``.
+The following (which would be used in the implementation) is wrong in stubs::
```

Comment:
I think this should be removed from the typing spec as well. This is more of a guide thing.